File: apps/mobile/android/app/src/main/java/br/com/aipinho/mobile/network/AgentApiClient.kt
package br.com.aipinho.mobile.network

import br.com.aipinho.mobile.models.AgentTabConfig
import br.com.aipinho.mobile.models.ConnectionProfile
import br.com.aipinho.mobile.ui.policies.ChatAutoRefreshPolicy
import java.net.URLEncoder

class AgentApiClient(
    profile: ConnectionProfile,
    tokenProvider: () -> String?,
    private val config: AgentTabConfig,
) : BaseApiClient(profile, profile.corePort, tokenProvider, ChatAutoRefreshPolicy.chatRequestTimeoutMs) {
    fun health() = get("${config.routePrefix}/health")

    fun configStatus() = get("${config.routePrefix}/config/status")

    fun sessions() = get("${config.routePrefix}/sessions")

    fun createSession(title: String = config.displayName) =
        post("${config.routePrefix}/sessions", "{\"title\":${JsonPayload.string(title)}}")

    fun renameSession(sessionId: String, title: String): ApiResponse {
        val path = "${config.routePrefix}/sessions/${encode(sessionId)}"
        return if (config.agentId == "lucio") {
            patch(path, "{\"title\":${JsonPayload.string(title)}}")
        } else {
            post("$path/rename", "{\"title\":${JsonPayload.string(title)}}")
        }
    }

    fun deleteSession(sessionId: String) =
        delete("${config.routePrefix}/sessions/${encode(sessionId)}")

    fun messages(sessionId: String) =
        get("${config.routePrefix}/sessions/${encode(sessionId)}/messages")

    fun send(
        sessionId: String,
        prompt: String,
        workspace: String,
        artifactIds: List<String>,
        autorun: Boolean,
        autoreview: Boolean,
        autoapproval: Boolean,
    ) = post(
        "${config.routePrefix}/sessions/${encode(sessionId)}/send",
        requestPayload(
            sessionId = sessionId,
            prompt = prompt,
            workspace = workspace,
            operationType = config.operationType,
            capabilities = emptyList(),
            artifactIds = artifactIds,
            autorun = autorun,
            autoreview = autoreview,
            autoapproval = autoapproval,
        ),
    )

    fun plan(sessionId: String, prompt: String, workspace: String) =
        post(
            "${config.routePrefix}/sessions/${encode(sessionId)}/plan",
            requestPayload(
                sessionId = sessionId,
                prompt = prompt,
                workspace = workspace,
                operationType = "${config.agentId}_plan",
                capabilities = if (workspace.isBlank()) emptyList() else listOf("read_workspace", "scan_workspace"),
            ),
        )

    fun preview(sessionId: String, prompt: String, workspace: String) =
        post(
            "${config.routePrefix}/sessions/${encode(sessionId)}/preview",
            requestPayload(
                sessionId = sessionId,
                prompt = prompt,
                workspace = workspace,
                operationType = "${config.agentId}_patch_preview",
                capabilities = listOf("read_workspace", "scan_workspace", "create_patch_preview"),
            ),
        )

    fun routePreview(sessionId: String, prompt: String, workspace: String, artifactIds: List<String>) =
        post(
            "${config.routePrefix}/sessions/${encode(sessionId)}/route-preview",
            requestPayload(
                sessionId = sessionId,
                prompt = prompt,
                workspace = workspace,
                operationType = config.operationType,
                capabilities = emptyList(),
                artifactIds = artifactIds,
            ),
        )

    fun viewModel(sessionId: String, afterEventId: String?, mode: String): ApiResponse {
        val query = buildList {
            afterEventId?.takeIf { it.isNotBlank() }?.let { add("after_event_id=${encode(it)}") }
            if (config.agentId == "lucio") add("mode=${encode(mode)}")
        }.joinToString("&")
        val suffix = query.takeIf { it.isNotBlank() }?.let { "?$it" }.orEmpty()
        return get("${config.routePrefix}/sessions/${encode(sessionId)}/view-model$suffix")
    }

    fun cancelRun(runId: String) =
        post("${config.routePrefix}/runs/${encode(runId)}/cancel")

    fun artifacts(sessionId: String) =
        get("/api/v1/artifacts/by-agent/${encode(config.agentId)}?session_id=${encode(sessionId)}")

    fun uploadTextArtifact(
        sessionId: String,
        filename: String,
        contentType: String,
        content: String,
        runId: String?,
        encoding: String = "text",
    ) = post(
        "/api/v1/agents/${encode(config.agentId)}/sessions/${encode(sessionId)}/artifacts/upload",
        "{" +
            "\"filename\":${JsonPayload.string(filename)}," +
            "\"content_type\":${JsonPayload.string(contentType)}," +
            "\"content\":${JsonPayload.string(content)}," +
            "\"encoding\":${JsonPayload.string(encoding)}," +
            "\"run_id\":${runId?.let(JsonPayload::string) ?: "null"}," +
            "\"origin\":\"mobile_upload\"" +
            "}",
    )

    private fun requestPayload(
        sessionId: String,
        prompt: String,
        workspace: String,
        operationType: String,
        capabilities: List<String>,
        artifactIds: List<String> = emptyList(),
        autorun: Boolean = true,
        autoreview: Boolean = true,
        autoapproval: Boolean = true,
    ): String {
        val caps = capabilities.joinToString(",") { JsonPayload.string(it) }
        val workspaceValue = workspace.takeIf { it.isNotBlank() }?.let(JsonPayload::string) ?: "null"
        val artifacts = artifactIds.joinToString(",") {
            "{\"artifact_id\":${JsonPayload.string(it)},\"purpose\":\"evidence\"}"
        }
        val common = mutableListOf(
            "\"session_id\":${JsonPayload.string(sessionId)}",
            "\"prompt\":${JsonPayload.string(prompt)}",
            "\"operation_type\":${JsonPayload.string(operationType)}",
            "\"requested_capabilities\":[$caps]",
        )
        when (config.agentId) {
            "lucio" -> {
                common.add("\"workspace_id\":$workspaceValue")
                common.add("\"artifacts\":[$artifacts]")
            }
            "codex_agent" -> {
                common.add("\"workspace_context\":$workspaceValue")
                common.add("\"autorun_enabled\":$autorun")
                common.add("\"autoreview_enabled\":$autoreview")
                common.add("\"autoapproval_enabled\":$autoapproval")
            }
            else -> common.add("\"workspace_context\":$workspaceValue")
        }
        return "{${common.joinToString(",")}}"
    }

    private fun encode(value: String): String = URLEncoder.encode(value, "UTF-8")
}

File: apps/launcher/ui/api/agent_api_client.py
from __future__ import annotations

from dataclasses import dataclass
from urllib.parse import quote

from apps.launcher.ui.api.base_client import ApiResult, BaseClient


@dataclass(frozen=True)
class AgentEndpointConfig:
    agent_id: str
    display_name: str
    route_prefix: str
    operation_type: str
    provider_label: str
    supports_plan: bool = True
    supports_preview: bool = True
    supports_route_preview: bool = False
    external_provider_notice: str | None = None


class DesktopAgentApiClient(BaseClient):
    def __init__(self, base_url: str, config: AgentEndpointConfig, **kwargs) -> None:
        super().__init__(base_url, **kwargs)
        self.config = config

    def health(self) -> ApiResult:
        return self.get(f"{self.config.route_prefix}/health")

    def config_status(self) -> ApiResult:
        return self.get(f"{self.config.route_prefix}/config/status")

    def sessions(self) -> ApiResult:
        return self.get(f"{self.config.route_prefix}/sessions")

    def create_session(self, title: str | None = None) -> ApiResult:
        return self.post(
            f"{self.config.route_prefix}/sessions",
            {"title": title or self.config.display_name},
        )

    def rename_session(self, session_id: str, title: str) -> ApiResult:
        path = f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}"
        if self.config.agent_id == "lucio":
            return self.patch(path, {"title": title})
        return self.post(f"{path}/rename", {"title": title})

    def delete_session(self, session_id: str) -> ApiResult:
        return self.delete(f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}")

    def messages(self, session_id: str) -> ApiResult:
        return self.get(f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}/messages")

    def send(
        self,
        session_id: str,
        prompt: str,
        workspace_context: str = "",
        artifact_ids: list[str] | None = None,
    ) -> ApiResult:
        return self.post(
            f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}/send",
            self._payload(
                session_id,
                prompt,
                workspace_context,
                self.config.operation_type,
                [],
                artifact_ids or [],
            ),
        )

    def plan(self, session_id: str, prompt: str, workspace_context: str = "") -> ApiResult:
        return self.post(
            f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}/plan",
            self._payload(
                session_id,
                prompt,
                workspace_context,
                f"{self.config.agent_id}_plan",
                ["read_workspace", "scan_workspace"] if workspace_context else [],
                [],
            ),
        )

    def preview(self, session_id: str, prompt: str, workspace_context: str = "") -> ApiResult:
        return self.post(
            f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}/preview",
            self._payload(
                session_id,
                prompt,
                workspace_context,
                f"{self.config.agent_id}_patch_preview",
                ["read_workspace", "scan_workspace", "create_patch_preview"],
                [],
            ),
        )

    def route_preview(
        self,
        session_id: str,
        prompt: str,
        workspace_context: str = "",
        artifact_ids: list[str] | None = None,
    ) -> ApiResult:
        return self.post(
            f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}/route-preview",
            self._payload(
                session_id,
                prompt,
                workspace_context,
                self.config.operation_type,
                [],
                artifact_ids or [],
            ),
        )

    def view_model(
        self,
        session_id: str,
        after_event_id: str | None = None,
        mode: str = "normal",
    ) -> ApiResult:
        query: list[str] = []
        if after_event_id:
            query.append(f"after_event_id={quote(after_event_id, safe='')}")
        if self.config.agent_id == "lucio":
            query.append(f"mode={quote(mode, safe='')}")
        suffix = f"?{'&'.join(query)}" if query else ""
        return self.get(
            f"{self.config.route_prefix}/sessions/{quote(session_id, safe='')}/view-model{suffix}"
        )

    def cancel_run(self, run_id: str) -> ApiResult:
        return self.post(f"{self.config.route_prefix}/runs/{quote(run_id, safe='')}/cancel")

    def artifacts(self, session_id: str) -> ApiResult:
        return self.get(
            f"/api/v1/agents/{quote(self.config.agent_id, safe='')}/sessions/"
            f"{quote(session_id, safe='')}/artifacts"
        )

    def _payload(
        self,
        session_id: str,
        prompt: str,
        workspace_context: str,
        operation_type: str,
        capabilities: list[str],
        artifact_ids: list[str],
    ) -> dict[str, object]:
        payload: dict[str, object] = {
            "session_id": session_id,
            "prompt": prompt,
            "operation_type": operation_type,
            "requested_capabilities": capabilities,
        }
        if self.config.agent_id == "lucio":
            payload["workspace_id"] = workspace_context or None
            payload["artifacts"] = [
                {"artifact_id": artifact_id, "purpose": "evidence"}
                for artifact_id in artifact_ids
            ]
        else:
            payload["workspace_context"] = workspace_context or None
        return payload
